### DISCUSSION TITLE
Class의 Computed Property에 관련하여 정리

### DISCUSSION THREAD

**Discussion by Chuncheonian** [swift-in-gangnam/swift-grammar#25] [Ideas]
# Computed Property

Class에서 `Computed property`라는 것을 왜 사용하는지를 가장 기초격인 C++를 통해 설명하도록 하겠습니다!

## C++
전체 C++ 코드는 다음과 같습니다.
```cpp
// c++
class Person {
  private:
    int age;
    string name;

  public:
    // 생성자 - swift의 init()이랑 같은 역할을 한다
    Person(string name, int age) {
      this->name = name;
      this->age = age;
    }
    
  int getAge() { return age; }  // age변수에 대한 getter
  
  string getName() { return name; }  // name변수에 대한 getter

  void setAge(int age) { this->age = age; } // age변수에 대한 setter

  void setName(string name) { this->name = name; }  // name변수에 대한 setter

};

Person p1("John", 20);

cout << p1.age << endl;  // age변수는 private 변수이므로 바로 접근이 불가 (Error!)
cout << p1.getAge() << endl;  // 이렇게 getter을 통해 접근 가능

```
이제, 하나하나씩 설명할게요!  

여기에서, private이란 class 내부에선 접근이 가능하지만, class 외부에선 바로 접근이 불가능하게 만들어 줍니다.

이를 통해, 변수를 함부로 변경하지 못하게 하여 객체 지향 프로그래밍에서 주요한 개념중 하나인 `은닉성`을 만족시킵니다.

```cpp
// c++
  private:
    int age;
    string name;
```

private하게 선언된 변수를 class 외부에서 접근하기 위해선 getter를 통해 접근이 가능
```cpp
// c++
cout << p1.age << endl;  // age변수는 private 변수이므로 바로 접근이 불가
cout << p1.getAge() << endl;  // 이렇게 getter을 통해 접근 가능
```
이처럼 대다수 언어에서 class에서는 getter, setter를 사용합니다.

## Swift

**이제, C++로 구현한 위의 코드를 Swift로 구현하겠습니다.**

```swift
// swift
class Person {
  private var age: Int
  private var name: String

  init(age: Int, name: String) {
    self.age = age
    self.name = name
  }

  var ageChange: Int {
    get { return age }
    set { self.age = newValue }
  }
}

var p1: Person = Person(age: 27, name: "dongyoung")
// print(p1.age)  // 저장 프로퍼티 age는 private이니까 class 외부에서 바로 접근이 불가!
print(p1.ageChange)  // 이런 식으로, getter을 통해 접근이 가능
p1.ageChange = 28
print(p1.ageChange)  // setter을 통해 pricate한 저장 프로퍼티 변경
```
Computed property는 실제 값을 저장하는 프로퍼티가 아니라, 특정 상태에 따른 값을 연산하는 프로퍼티입니다.

Person 클래스에는 두 개의 저장 프로퍼티가 있습니다. 이 때 private으로 하여 Class 외부에서 직접적으로 접근하지 못하도록 하였습니다.

이를 통해, 은닉성을 만족하도록 하였습니다. 대부분 객체 지향 프로그래밍에서는 이 개념을 만족하도록 코드를 짭니다.

```swift
// swift
  private var age: Int
  private var name: String
```

private하게 선언된 Stored Property를 Class 외부에선 ageChange라는 Computed Property를 통해서만 getter, setter를 할 수 있습니다.

```swift
// swift
var p1: Person = Person(age: 27, name: "dongyoung")
// print(p1.age)  // 저장 프로퍼티 age는 private이니까 class 외부에서 바로 접근이 불가!
print(p1.ageChange)  // 이런 식으로, getter을 통해 접근이 가능
```

## 결론
다시 한번 정리하자면,

1. 은닉성을 위해 Stored Property는 private하게 선언하기 때문에, Class 외부에서 Stored Property를 접근하기 위해 getter & setter를 사용한다.
물론, Stored Property를 private하게 선언 안하면, getter & setter를 사용할 필요가 없다.
(이것의 설명이 책에서 생략됨)
```swift
// Swift
class Person {
  var age: Int  // private X
  private var name: String  // private

  var nameComputed: String {
    get { return name }
    set { self.name = newValue }
  }
  
}

var p1: Person = Person(age: 27, name: "John")
print(p1.age)  // age는 직접적 출력 가능
print(p1.name)  // name은 private하므로 직접적 출력 불가

print(p1.nameComputed)  // name은 Computed Property 통해 출력 가능
```

2. Swift는 C++에서의 두개의 메소드를 하나의 Computed property 만으로 getter, setter를 구현하도록 하였다.

```swift
// C++에서는 두 개의 함수 `getAge()`, `setAge()`를 통해 getter와 setter를 구현
int getAge() { return age; }  // age변수에 대한 getter
void setAge(int age) { this->age = age; }  // age변수에 대한 setter

// Swift에서는 하나의 Computed Property `ageChange` 를 통해 getter와 setter를 구현
var ageChange: Int {
    get { return age }
    set { self.age = newValue }
  }
```

## 참고

책에서는 private, public가 사용된 Class 설명이 안되어 있어서 Computed Property를 공부할 때 많이 헷갈렸습니다.

그래서 클래스의 가장 기초격인 C++를 통해 개념을 이해를 하여 이렇게 글을 남겼습니다!

항상 객체 지향 프로그래밍은 은닉성, 캡슐화, 추상화, 다형성 등 특징을 가지고 있다는 것을 상기를 하셔야 합니다.

혹시 객체 지향 프로그래밍 개념이 헷갈리시면 [유튜브](https://www.youtube.com/watch?v=vrhIxBWSJ04&t=500s)를 보시면 이해가 잘 되실 거에요!

**Comment by iammiori** [comment]
제가 찐 컴공은 아니라 c++ 에 약한데 좋은 정리 감사합니다 :)
